File: gitterscrubber.js
'use strict';

const http = require('http');
const request = require('request');
const url = require('url');
const querystring = require('querystring');

// your Gitter oauth key
const key = 'put your gitter oauth key here';
// your Gitter oauth secret
const secret = 'put your gitter oauth secret here';
// your Gitter redirect aka the address of this server
const redirect = 'http://global address to this server:port/';
// port of this server
const port = 3000;

const scrubOneToOnes = true;

// click this link with ctrl + mouse click to authenticate in browser
console.log('Click this to scrub: ' + 'https://gitter.im/login/oauth/authorize?' + querystring.stringify({
  client_id: key,
  response_type: 'code',
  redirect_uri: redirect
}));

function fetchMessages(token, roomId, beforeId, cb) {
  var buffer = '';
  request.get({url: 'https://api.gitter.im/v1/rooms/' + roomId + '/chatMessages?limit=100' + (beforeId ? ('&beforeId=' + beforeId) : ''), headers: {
    'authorization': 'Bearer ' + token,
    'accept': 'application/json'
  }}).on('data', function(data) {
    buffer += data.toString();
  }).on('end', function() {
    var messages = JSON.parse(buffer);
    if (messages.error) {
      console.log('Failed, retrying in a while!');
      console.log(messages.error);
      setTimeout(function() {
        fetchMessages(token, roomId, beforeId, cb);
      }, 1000);
    } else {
      cb(messages);
    }
  });
}

function deleteMessage(token, roomId, messageId, cb) {
  request.delete({url: 'https://api.gitter.im/v1/rooms/' + roomId + '/chatMessages/' + messageId, headers: {
    'authorization': 'Bearer ' + token
  }}).on('response', function(response) {
    if (response.statusCode != 204) {
      console.log('Failed to delete, retrying in a while!');
      setTimeout(function() {
        deleteMessage(token, roomId, messageId, cb);
      }, 1000);
    } else {
      cb();
    }
  });
}

function deleteMyMessages(token, userId, roomId, messages, index, cb) {
  if (index == messages.length) {
    cb();
    return;
  }
  if (messages[index].fromUser.id === userId) {
    deleteMessage(token, roomId, messages[index].id, function() {
        console.log('Deleted message: ' + messages[index].text);
        deleteMyMessages(token, userId, roomId, messages, index + 1, cb);
    });
  } else {
    deleteMyMessages(token, userId, roomId, messages, index + 1, cb);
  }
}

function readAllMessages(token, userId, roomId, beforeId, cb) {
  fetchMessages(token, roomId, beforeId, function(messages) {
    if (messages.length) {
      console.log('Meddelanden: ' + messages.length + ' Första datum: ' + messages[0].sent);
    }

    deleteMyMessages(token, userId, roomId, messages, 0, function() {
      if (messages.length) {
        readAllMessages(token, userId, roomId, messages[0].id, cb);
      } else {
        console.log('Reached the end of this room, calling back!');
        cb();
      }
    });
  });
}

function clearRooms(token, userId, rooms, index) {
  if (index == rooms.length) {
    return;
  }
  // scrub everything, even one to one
  if (!scrubOneToOnes && rooms[index].oneToOne) {
    clearRooms(token, userId, rooms, index + 1);
  } else {
    console.log('Scrubbing all messages in room: ' + rooms[index].name);
    readAllMessages(token, userId, rooms[index].id, null, function() {
      clearRooms(token, userId, rooms, index + 1);
    });
  }
}

function getAccessToken(code, cb) {
  var buffer = '';
  request.post({url: 'https://gitter.im/login/oauth/token', body: JSON.stringify({
    client_id: key,
    client_secret: secret,
    redirect_uri: redirect,
    grant_type: 'authorization_code',
    code: code
  }), headers: {
    'content-type': 'application/json',
    'accept': 'application/json'
  }})
  .on('data', function(data) {
    buffer += data.toString();
  }).on('end', function() {
    var access = JSON.parse(buffer);
    cb(access.access_token);
  });
}

function getUser(token, cb) {
  var buffer = '';
  request.get({url: 'https://api.gitter.im/v1/user/me', headers: {
    'authorization': 'Bearer ' + token,
    'accept': 'application/json'
  }}).on('data', function(data) {
    buffer += data.toString();
  }).on('end', function() {
    var user = JSON.parse(buffer);
    cb(user.id, user.displayName);
  });
}

function getPublicRooms(token, cb) {
  var buffer = '';
  request.get({url: 'https://api.gitter.im/v1/user/me/rooms', headers: {
    'authorization': 'Bearer ' + token,
    'accept': 'application/json'
  }}).on('data', function(data) {
    buffer += data.toString();
  }).on('end', function() {
    var rooms = JSON.parse(buffer);
    cb(rooms);
  });
}

const server = http.createServer((req, res) => {
  var code = querystring.parse(url.parse(req.url).query).code; // error om deny
  if (code && code.length == 40) {
    getAccessToken(code, function(token) {
      getUser(token, function(userId, userName) {
        console.log('[Scrubbing user: ' + userName + ']');
        getPublicRooms(token, function(rooms) {
          // respond with what rooms will be deleted and who you are
          var response = '';
          response += '<h1>gitterscrubber.org</h1>';
          response += 'You are logged in as: <b>' + userName + '</b>';
          response += '<h3>Deleting your messages in public rooms:</h3>';
          for (let i = 0; i < rooms.length; i++) {
            // delete everything
            if (scrubOneToOnes || !rooms[i].oneToOne) {
              response += '<b>' + rooms[i].name + '</b><br>';
            }
          }
          response += '<p>Your task has been posted. This will take a lot of time to finish. Expect at least a day of waiting.</p>';
          res.end(response);
          // start the scrubbing
          clearRooms(token, userId, rooms, 0);
        });
      });
    });
  } else {
    res.end('Seems like you denied access');
  }
});

server.listen(port);

File: gitterscrubber.go
// gitterscrubber.org
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
)

const (
	// your Gitter oauth key
	key = ""
	// your Gitter oauth secret
	secret = ""
	// your Gitter redirect aka the address of this server
	redirect = ""
	// port of this server
	port = 3000
	// scrub priavet 1 to 1 rooms also?
	scrubOneToOnes = true
)

// Room is a Gitter room
type Room struct {
	Name     string
	OneToOne bool
	ID       string
}

// User is a Gitter user
type User struct {
	ID          string
	DisplayName string
}

// Message is a Gitter message
type Message struct {
	Sent     string
	ID       string
	FromUser User
	Text     string
}

func getAccessToken(code string) string {

	values := map[string]string{
		"client_id":     key,
		"client_secret": secret,
		"redirect_uri":  redirect,
		"grant_type":    "authorization_code",
		"code":          code}

	jsonValue, _ := json.Marshal(values)

	resp, err := http.Post("https://gitter.im/login/oauth/token", "application/json", bytes.NewBuffer(jsonValue))

	if err != nil {

	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)

	if err == nil {
		var tmp map[string]string
		json.Unmarshal(body, &tmp)
		return tmp["access_token"]
	}
	fmt.Println("Error: getAccessToken")
	return ""
}

func gitterRequest(method string, expectedStatusCode int, url, token string) ([]byte, error) {
	client := http.Client{}
	request, _ := http.NewRequest(method, url, nil)
	request.Header.Add("authorization", ("Bearer " + token))
	resp, err := client.Do(request)

	if err != nil {
		fmt.Println("Error: gitterRequest")
		return []byte{}, err
	}

	defer resp.Body.Close()

	if resp.StatusCode != expectedStatusCode {
		fmt.Println("Error: gitterRequest (" + resp.Status + ")")
		//time.Sleep(500 * time.Millisecond)
		//return gitterRequest(method, expectedStatusCode, url, token)
	}

	return ioutil.ReadAll(resp.Body)
}

func getUser(token string) (string, string) {
	body, err := gitterRequest("GET", 200, "https://api.gitter.im/v1/user/me", token)

	if err != nil {
		fmt.Println("Error: getUser")
		return "", ""
	}

	var user User
	json.Unmarshal(body, &user)
	return user.ID, user.DisplayName
}

func getPublicRooms(token string) []Room {
	body, err := gitterRequest("GET", 200, "https://api.gitter.im/v1/user/me/rooms", token)

	if err == nil {
		var rooms []Room
		json.Unmarshal(body, &rooms)
		return rooms
	}
	fmt.Println("Error: getPublicRooms")
	return []Room{}
}

func fetchMessages(token string, roomID string, beforeID string) []Message {

	if len(beforeID) > 0 {
		beforeID = "&beforeId=" + beforeID
	}

	body, err := gitterRequest("GET", 200, ("https://api.gitter.im/v1/rooms/" + roomID + "/chatMessages?limit=100" + beforeID), token)

	var messages []Message
	if err == nil {
		json.Unmarshal(body, &messages)
	}
	return messages
}

func deleteMyMessages(token string, userID string, roomID string, messages []Message, index int) {
	if index == len(messages) {
		return
	}
	if messages[index].FromUser.ID == userID {
		body, err := gitterRequest("DELETE", 204, ("https://api.gitter.im/v1/rooms/" + roomID + "/chatMessages/" + messages[index].ID), token)
		if err == nil {
			var tmp []Message
			json.Unmarshal(body, &tmp)
			return
		}
		fmt.Println("Deleted message: " + messages[index].Text)
	}
	deleteMyMessages(token, userID, roomID, messages, (index + 1))
}

// top most function for scrubbing a room
func scrubRoom(token string, userID string, roomID string, beforeID string) {
	// zero messages means out of messages, or error
	messages := fetchMessages(token, roomID, beforeID)
	deleteMyMessages(token, userID, roomID, messages, 0)
	if len(messages) == 0 {
		return
	}
	scrubRoom(token, userID, roomID, messages[0].ID)
}

func main() {
	v := url.Values{}
	v.Add("client_id", key)
	v.Add("response_type", "code")
	v.Add("redirect_uri", redirect)

	// click this link with ctrl + mouse click to authenticate in browser
	fmt.Println("Click this to scrub: " + "https://gitter.im/login/oauth/authorize?" + v.Encode())

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		code := r.URL.Query().Get("code")
		if len(code) == 40 {
			token := getAccessToken(code)
			userID, userName := getUser(token)
			fmt.Println("[Scrubbing user: " + userName + "]")
			rooms := getPublicRooms(token)

			response := "<h1>gitterscrubber.org</h1>"
			response += "You are logged in as: <b>" + userName + "</b>"
			response += "<h3>Deleting your messages in public rooms:</h3>"

			for _, room := range rooms {
				if scrubOneToOnes || !room.OneToOne {
					response += "<b>" + room.Name + "</b><br>"
				}
			}
			response += "<p>Your task has been posted. This will take a lot of time to finish. Expect at least a day of waiting.</p>"
			fmt.Fprint(w, response)

			go func() {
				for _, room := range rooms {
					if !(!scrubOneToOnes && room.OneToOne) {
						fmt.Println("Scrubbing all messages in room: " + room.Name)
						scrubRoom(token, userID, room.ID, "")
					}
				}
				fmt.Println("Done scrubbing user: " + userName)
			}()
		}
	})
	fmt.Print(http.ListenAndServe("0.0.0.0:3000", nil))
}
